medcat/utils/relation_extraction/rel_dataset.py

```diff
@@ -0,0 +1,503 @@
+                                else:
+                                    left_context_start_char_pos = tokenizer_data[
+                                        "offset_mapping"][ent1_left_ent_context_token_pos_end][0]
+
+                                ent2_right_ent_context_token_pos_end = ent2_token_start_pos + self.ent_context_right
+                                if ent2_right_ent_context_token_pos_end >= doc_length_tokens - 1:
+                                    ent2_right_ent_context_token_pos_end = doc_length_tokens - 2
+                                else:
+                                    right_context_start_end_pos = tokenizer_data[
+                                        "offset_mapping"][ent2_right_ent_context_token_pos_end][1]
+
+                                window_tokenizer_data = self.tokenizer(
+                                    text[left_context_start_char_pos:right_context_start_end_pos])
+                                # update token loc to match new selection
+                                ent2_token_start_pos = ent2_token_start_pos - ent1_token_start_pos
+                                ent1_token_start_pos = self.ent_context_left if ent1_token_start_pos - \
+                                    self.ent_context_left > 0 else ent1_token_start_pos
+                                ent2_token_start_pos += ent1_token_start_pos
+                                ent1_ent2_new_start = (
+                                    ent1_token_start_pos, ent2_token_start_pos)
+                                en1_start, en1_end = window_tokenizer_data[
+                                    "offset_mapping"][ent1_token_start_pos]
+                                en2_start, en2_end = window_tokenizer_data[
+                                    "offset_mapping"][ent2_token_start_pos]
+
+                                relation_instances.append([window_tokenizer_data["input_ids"], ent1_ent2_new_start, start_entity_value, end_entity_value, relation_label, self.blank_label_id,
+                                                           start_entity_types, end_entity_types, start_entity_id, end_entity_id, start_entity_cui, end_entity_cui, doc_id, "",
+                                                           en1_start, en1_end, en2_start, en2_end])
+
+                    output_relations.extend(relation_instances)
+
+        all_relation_labels = [relation[4] for relation in output_relations]
+
+        nclasses, labels2idx, idx2label = self.get_labels(
+            all_relation_labels, self.config)
+
+        # replace label_id with actual detected label number
+        for idx in range(len(output_relations)):
+            output_relations[idx][5] = labels2idx[output_relations[idx][4]]
+
+        return {"output_relations": output_relations, "nclasses": nclasses, "labels2idx": labels2idx, "idx2label": idx2label}
+
+    @classmethod
+    def get_labels(cls, relation_labels: List[str], config: ConfigRelCAT) -> Any:
```

Comment:
The return type seems to be a `Tuple[int, Dict, Dict]`.
A doc string would also be useful.